### ISSUE TITLE
Links to documentation don't work in VSCode

### ISSUE THREAD

**Issue by bkonyi** [flutter/devtools#6105]
Clicking on links to documentation doesn't open the link in a browser when using embedded DevTools tooling in VSCode.

FYI @DanTup

**Comment by DanTup** [comment]
Seems to be a VS Code restriction of the iframe sandbox. There's an issue about this (please add 👍 ) here: https://github.com/microsoft/vscode/issues/186043

**Comment by DanTup** [comment]
Looks like VS Code isn't going to add support for this:

https://github.com/microsoft/vscode/issues/186043#issuecomment-1744900853

So I think the only thing we can do is pass some flag in the querystring (or call some function with `postMessage`) to tell DevTools that any links that need to open in new pages need to instead call an API (via postMessage) to ask the VS Code extension to open the browser windows.

This feels really ugly (and is very VS Code-specific), but I can't think of any simpler option.

@kenzieschmoll any thoughts or other ideas?

**Comment by kenzieschmoll** [comment]
We should already have the `ide` and `embedded` query param we can look at to see if we are embedded in VS code. As for DevTools controlling whether we launch in the browser or not, I think we can do that by passing `launchMode: LaunchMode.externalApplication` to our call to `launchUrl` from `package:url_launcher`. We'd want to pass this when we are in embeddedMode + open from VS Code.

**Comment by DanTup** [comment]
This seems to work for me today, although I'm not sure why. It still writes an error to the console, but the page opens fine.

@bkonyi do you still see it not work?


https://github.com/flutter/devtools/assets/1078012/e296509a-8703-4d89-a212-f70a1e9db04c

**Comment by bkonyi** [comment]
Seems to work on my end now. That's strange... I think we can close this for now.

**Comment by kenzieschmoll** [comment]
Re-opening since we had another report of this issue here: https://github.com/flutter/devtools/issues/6557

**Comment by DanTup** [comment]
Ok, I figured out why I saw it once but not now. It seems to fail on current Flutter stable, but it works fine on master and beta.

My suspicion is that something has changed in Flutter (or DevTools) since last stable that has fixed this (although I'm not sure what).

Can anyone else that sees this test using Flutter master or beta and confirm it's fixed for them too? (@osaxma)

Thanks!

**Comment by CoderDake** [comment]
I did some testing on this today and notice that this is broken on flutter stable, but works on flutter beta. So perhaps this has fixed on our end somehow 🤔

**Comment by CoderDake** [comment]
@DanTup looks like I confirmed in parallel to you and we posted at the same time 😆 🎉

**Comment by osaxma** [comment]
> Can anyone else that sees this test using Flutter master or beta and confirm it's fixed for them too?

@DanTup: it doesn't work for me in the latest master on MacOS:

https://github.com/flutter/devtools/assets/46427323/ed157551-c12b-4b02-929a-874be3c6956e



```
Flutter 3.16.0-14.0.pre.23 • channel master • https://github.com/flutter/flutter.git
Framework • revision eefafdacdb (3 hours ago) • 2023-10-20 00:06:34 -0400
Engine • revision 9dd59f7fcf
Tools • Dart 3.3.0 (build 3.3.0-45.0.dev) • DevTools 2.28.1
---
Devtools 2.25.0
Dart VS Code extension: v3.74.0
VS Code version: 1.83.1
```

**Comment by DanTup** [comment]
@osaxma the video above shows DevTools 2.25.0 which is not the latest from master (the version output you posted shows "DevTools 2.28.1").

It seems like maybe either VS Code is not using the version of the SDK you expect (you can check this from the `{}` in the status bar next to Dart), or you're somehow getting a cached version of DevTools loaded into the webview.

Can you verify the version in the status bar and restart VS Code and see whether you still see v2.25.0 showing up?

**Comment by osaxma** [comment]
> It seems like maybe either VS Code is not using the version of the SDK you expect (you can check this from the `{}` in the status bar next to Dart)

@DanTup:
Oops! Nice catch and you're totally right.. VS Code was not using the version in path for some reason (which I change using the `fvm` version management tool).  Below `~/fvm/default` is a symlink to the selected version. 

<img width="623" alt="Screen Shot 2023-10-23 at 1 15 31 PM" src="https://github.com/flutter/devtools/assets/46427323/6ab07dd0-620e-464d-82d9-659415d7002b">

----


Anyway, I can confirm that it's working in `Flutter 3.16.0-14.0.pre.23 • channel master`

_Sorry for any inconvenience -- time to go update some Flutter issues where I confidently claimed the issue was reproducible in master 😅_

**Comment by DanTup** [comment]
Great, thanks for confirming. I'm going to close this on the basis that several people have seen it broken in current stable but fixed in newer Flutters. If anyone finds differently (and confirms the SDK version as discussed above), we can re-open.


> VS Code was not using the version in path for some reason

If you think VS Code is not using the version you'd _expect_, it's worth filing a Dart-Code issue (or at least reviewing the logs by running **Dart; Open Extension Logs**). There are some reasons we might not use the one you expect (for example your `.dart_tool/package_config.json` references a different one to `PATH`/settings, we'll still use that because it's the one the analyzer is going to be using).

**Comment by DanTup** [comment]
I'm not sure why this was broken in stable, but while reviewing another of @CoderDake's changes I realised that this was fixed by calling up to VS Code by @tomgilder some time back:

https://github.com/flutter/devtools/pull/3252

Which explains why other VS Code extension authors are still complaining about this even though it seemed fixed to us 🙃

**Comment by kenzieschmoll** [comment]
@DanTup can you confirm whether this issue is still happening on the Flutter stable that was shipped today? (flutter 3.16)

**Comment by DanTup** [comment]
Yep, it previously worked for me on `beta` and it works today on `stable`. It was only the previous stable that was broke (for reasons I haven't investigated).

**Comment by DanTup** [comment]
Oops, I inverted my response.

The issue is resolved in the release that went stable yesterday. It previously worked for me on `beta` and I tested just now on `stable`.

**Comment by kenzieschmoll** [comment]
Do we know what the issue / fix was? between the previous stable and the one that went out yesterday?

**Comment by DanTup** [comment]
I don't know - I didn't dig into it much because it seemed fixed. I think it's worth figuring out though (to ensure it can't accidentally break in the same way in future). I'll see what I can find.

**Comment by DanTup** [comment]
I went back to previous stable and I see this error:

![image](https://github.com/flutter/devtools/assets/1078012/b4af71af-a42b-40fb-b2ea-fdd534fe4c66)

There's an open issue about this at https://github.com/flutter/flutter/issues/133023 but it's not clear to me why it would have changed.

The code in DevTools [calls `url_launcher.launchUrl` and then `launchUrlVSCode`](https://github.com/flutter/devtools/blob/2db07727f0cdc9e8da6e246dabc91a5265d59546/packages/devtools_app/lib/src/shared/config_specific/launch_url/launch_url.dart#L23). If the first call (url_launcher) throws, we'll never call the VS Code function. The expected behaviour in VS Code was for `url_launcher` to fail silently, so this was fine.

So, to avoid this issue in future (since it's not clear what the cause of https://github.com/flutter/flutter/issues/133023 is or if it's likely to come back) I suggest that we add a try/catch and also call `launchUrlVsCode` in the case of an exception. This isn't risky because all it does is called `postMessage` with the URL to open, so outside of VS Code nobody will be listening anyway.

@kenzieschmoll @CoderDake thoughts?

**Comment by CoderDake** [comment]
@DanTup that sounds reasonable to me. I've opened up an issue where we can keep track of that item :)
https://github.com/flutter/devtools/issues/6778 . Perhaps @kenzieschmoll can weigh in on how we want to prioritize/proceed with it :)

**Comment by kenzieschmoll** [comment]
adding the try catch sgtm. Let's prioritize this ASAP and we can try to put it into the cherry-pick release here: https://github.com/dart-lang/sdk/issues/54085
